### PR TITLE
storage updates

### LINKED ISSUE #102: LocalStorage - datetime storage
localstorage should be able to handle datetime objects

best to wait until strptime gets added to skulpt before attempting this

### LINKED ISSUE #176: storage - should fail loudly in serialization
i.e. if we get something we can't json or isn't bytes.

### PR DESCRIPTION
**Pull request by s-cork**
close #176 - add a serializer to fail loudly
close #102 - support serializing datetimes

fix a bug with iterating over a storage object.
`__iter__` can't suspend.

### CODE REVIEW

File: client_code/storage.py [hunk 1]
```diff
@@ -19,22 +21,64 @@
...
+
+def _serialize(obj):
+    # we won't support subclasses of builtins so just check type is
+    ob_type = type(obj)
+    if ob_type is list or ob_type is tuple:
+        ret = []
+        for item in obj:
+            ret.append(_serialize(item))
+        return ret
+    elif ob_type is dict:
```
Comment (meatballs) [inline client_code/storage.py:37]: How about:
```
elif ob_type is dict:
    assert(all([type[k] is str for k in obj])
    return {k, _serialize(v) for k, v in obj.items()

File: client_code/storage.py [hunk 2]
```diff
@@ -19,22 +21,64 @@
...
+    elif ob_type is datetime:
+        return {_SPECIAL + "datetime": obj.isoformat()}
+    elif ob_type is date:
+        return {_SPECIAL + "date": obj.isoformat()}
+    else:
+        raise TypeError(f"Cannot serialize an object of type {ob_type.__name__}")
+
+
+def _special_deserialize(key, value):
+    key = key[len(_SPECIAL) :]
```
Comment (meatballs) [inline client_code/storage.py:56]: How about:
```
deserializers = {"date": date.fromisoformat, "datetime": datetime.fromisoformat}
try:
    return deserializers[key](value)
except KeyError:
    raise ValueError(f"unknown special deserialization type {key!r}")
```

File: client_code/storage.py [hunk 3]
```diff
@@ -19,22 +21,64 @@
...
+        return datetime.fromisoformat(value)
+    raise ValueError(f"unknown special deserialization type {key!r}")
 
 
 def _deserialize(obj):
     """convert simple proxy objects (and nested simple proxy objects) to dictionaries"""
-    # @TODO datetime objects - we'd also need a _serialize method for that
-    if isinstance(obj, list):
+    ob_type = type(obj)
+    if ob_type is list:
```
Comment (meatballs) [inline client_code/storage.py:67]:
```
return [_deserialize(item) for item in obj]
```

File: client_code/storage.py [hunk 4]
```diff
@@ -19,22 +21,64 @@
...
         # Then we're a simple proxy object
         # keys are strings so only _deserialize the values
         ret = {}
         # use _Object.keys to avoid possible name conflict
-        for key in _Object.keys(obj):
+        keys = _Object.keys(obj)
+        if len(keys) == 1 and keys[0].startswith(_SPECIAL):
+            key = keys[0]
+            return _special_deserialize(key, obj[key])
+        for key in keys:
```
[the remark this answers is not in the document]
Comment (s-cork) [reply client_code/storage.py:81]: Yeap agree with all of these - clearly too much javascript recently

**Pull request merged**